examples/server/webui/src/components/useAutosizeTextarea.ts

```diff
@@ -0,0 +1,68 @@
+import { useEffect, useRef, useState, useCallback } from 'react';
+
+const adjustTextareaHeight = (textarea: HTMLTextAreaElement | null) => {
+  if (!textarea) {
+    return;
+  }
+
+  const computedStyle = window.getComputedStyle(textarea);
+  const currentMaxHeight = computedStyle.maxHeight;
+
+  textarea.style.maxHeight = 'none';
+  textarea.style.height = 'auto';  
+
+  const scrollH = textarea.scrollHeight;  
+
+  textarea.style.height = `${scrollH}px`;  
+  textarea.style.maxHeight = currentMaxHeight;  
+};
+
+export interface AutosizeTextareaApi {
+  value: () => string;
+  setValue: (value: string) => void;
+  focus: () => void;
+  ref: React.RefObject<HTMLTextAreaElement>;
+  onInput: (event: React.FormEvent<HTMLTextAreaElement>) => void;  
+}
+
+export function useAutosizeTextarea(initValue: string): AutosizeTextareaApi {

```

Comment:
Now that this hook has 2 purposes (auto-resize + optimize rendering), I think we should have a more generic name for it, maybe `useChatTextarea`